### PR TITLE
MGMT-1917 Make pull secret editable in OCM

### PR DESCRIPTION
**Pull request by mareklibra**
With this patch, the OCM user can edit pull-secret. Since this is expected to be a rare case, the field is not visible by default. Prior this patch, this field was hidden in OCM.

The pull-secret is not visible by default (is collapsed) but the user canexpand an area to edit it.
If pull-secret automatic fetch fails, the area is expanded by default (plus error message is shown).

Outside of the OCM, the pull-secret filed is always rendered.

### CODE REVIEW

**Comment by mareklibra** [comment]
---

**Comment by mareklibra** [comment]
Cc @andybraren

**Comment by andybraren** [comment]
@mareklibra I've added a new design iteration to [MGMT-1917](https://issues.redhat.com/browse/MGMT-1917), let me know what you think. Regardless, this may be a good time to add the `?` help text to describe what pull secrets are/do as well. The "Learn more..." link would direct to `https://cloud.redhat.com/openshift/install/pull-secret` and we can enhance the description on that page via [PD-645](https://issues.redhat.com/browse/PD-645).

**Comment by mareklibra** [comment]
**Stand-alone:**


---

**In OCM:**


---



---

File: src/components/clusters/PullSecret.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        <Checkbox
+          id="checkbox-pull-secret"
+          label={
+            <span onClick={(event) => event.preventDefault()}>
+              Customize pull secret <PullSecretInfo />
+            </span>
+          }
+          aria-label="customize pull secret"
+          isChecked={isExpanded}
+          onChange={() => setExpanded(!isExpanded)}
```
Comment (jtomasek) [inline src/components/clusters/PullSecret.tsx:77]: I think we need to reset the `pullSecret` value to initial value when user disables the checkbox.

**Review by jtomasek — CHANGES_REQUESTED** [review]
`src/components/clusters/PullSecret.tsx` probably better fits in `src/components/clusterConfiguration` dir.

File: src/components/clusters/PullSecret.tsx (hunk 1 above)
Comment (mareklibra) [reply src/components/clusters/PullSecret.tsx:77]: good catch, it's not just hiding but "Customize" now

**Comment by nirfarkas** [comment]
Wording nit:
I think "Edit pull secret" is better here. @andybraren ?

File: src/components/clusters/PullSecret.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import { ClusterCreateParams, ocmClient } from '../../api';
+import { CLUSTER_MANAGER_SITE_LINK, PULL_SECRET_INFO_LINK } from '../../config';
+import { TextAreaField } from '../ui';
+import { useFormikContext } from 'formik';
+
+export type PullSecretProps = {
+  pullSecret?: string;
+};
+
+const ClusterManagerSiteLink = ({ text = 'OpenShiftCluster Manager' }: { text?: string }) => (
```
Comment (andybraren) [inline src/components/clusters/PullSecret.tsx:13]:
```suggestion
const ClusterManagerSiteLink = ({ text = 'OpenShift Cluster Manager' }: { text?: string }) => (
```

File: src/components/clusters/PullSecret.tsx [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      setExpanded(!isExpanded);
+    };
+
+    return (
+      <>
+        <Checkbox
+          id="checkbox-pull-secret"
+          label={
+            <span onClick={(event) => event.preventDefault()}>
+              Customize pull secret <PullSecretInfo />
```
Comment (andybraren) [inline src/components/clusters/PullSecret.tsx:83]:
```suggestion
              Edit pull secret <PullSecretInfo />
```

I think Nir is right - `Edit` is simpler and more common.

File: src/components/clusters/PullSecret.tsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            </span>
+          }
+          aria-label="customize pull secret"
```
Comment (andybraren) [inline src/components/clusters/PullSecret.tsx:86]:
```suggestion
          aria-label="edit pull secret"
```

**Review by andybraren — COMMENTED** [review]
Nit: I'm not sure how to edit the code for this, but the `?` popover for c.rh.c should have a link for "Learn more about pull secrets" instead of "in OpenShift Cluster Manager" since the user is already in OCM.

image

File: src/components/clusters/PullSecret.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  </>
+);
+
+const PullSecretInfo = () => (
+  <Popover
+    position="top"
+    bodyContent={
+      <>
+        Pull secrets are used to download OpenShift Container Platform components and connect
+        clusters to a Red Hat account.
```
Comment (jtomasek) [inline src/components/clusters/PullSecret.tsx:43]:
```suggestion
        clusters to a Red Hat account.{' '}
```

File: src/components/clusters/PullSecret.tsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        ) : (
+          <>
+            Pull secrets can be found in&nbsp;
+            <ClusterManagerSiteLink />
+          </>
+        )}
+      </>
+    }
+  >
+    <OutlinedQuestionCircleIcon />
```
Comment (jtomasek) [inline src/components/clusters/PullSecret.tsx:55]:
```suggestion
    <button onClick={(e) => e.preventDefault()} className="pf-c-form__group-label-help">
      <OutlinedQuestionCircleIcon noVerticalAlign />
    </button>
```
(according to https://www.patternfly.org/v4/components/form#basic)

We could eventually extract this into a specific `<FormFieldHelpPopover>` component.

File: src/components/clusters/PullSecret.tsx [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            aria-label="edit pull secret"
+            isChecked={isExpanded}
+            onChange={onCheckboxChange}
+            label="Edit pull secret"
+          />
+        }
+        labelIcon={<PullSecretInfo />}
+      >
+        {isExpanded && textArea}
+      </FormGroup>
```
Comment (jtomasek) [inline src/components/clusters/PullSecret.tsx:104]:
```suggestion
      <>
        <Checkbox
          id="checkbox-pull-secret"
          className="pf-u-display-inline-flex"
          aria-label="edit pull secret"
          isChecked={isExpanded}
          onChange={onCheckboxChange}
          label={
            <>
              Edit pull secret <PullSecretInfo />
            </>
          }
        />
        {isExpanded && textArea}
      </>
```

Standalone checkbox seems to work fine unless I am missing something.

**Review by jtomasek — CHANGES_REQUESTED** [review]

File: src/components/clusters/PullSecret.tsx (hunk 6 above)
Comment (jtomasek) [reply src/components/clusters/PullSecret.tsx:55]: This change ensures correct icon size and alignment.

File: src/components/clusters/PullSecret.tsx (hunk 7 above)
Comment (mareklibra) [reply src/components/clusters/PullSecret.tsx:104]: The `FormGroup` was added in attempt to fix alignment and event propagation. It's a good idea to wrap the icon at lower level, so I am reverting this change here.

**Review by jtomasek — APPROVED** [review]

**Pull request merged**